### PR TITLE
Add `Storage` interface wiki article

### LINKED ISSUE #816: [Doc] Add `Storage` adapter wiki article
## Description

Add content to the wiki article that explains the `Storage` interface: for whom it may be interesting to implement and the relevant technical details to do so.

## Reference material

Wiki article: https://wiki.iota.org/identity.rs/advanced/storage_adapter

## To-do List
Create a task-specific to-do list 

- [ ] Add content to the `Storage` wiki article

### PR DESCRIPTION
**Pull request by PhilippGackstatter**
# Description of change

Add `Storage` interface wiki article. This embeds the Rust and TS `MemStore` examples, which were documented, so they are more helpful for readers looking for an example implementation.

This also gets rid of the `index` in the Rust `MemStore`, where the `vaults` can be used as an implicit index, just like in the TS MemStore already.

## Links to any relevant issues

fixes #816

## Type of change
Add an `x` to the boxes that are relevant to your changes.

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] Enhancement (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Fix

## How the change has been tested

n/a

## Change checklist
Add an `x` to the boxes that are relevant to your changes.

- [x] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by cycraig** [comment]
Change the target branch to `support/v0.5`.

File: documentation/docs/advanced/storage_interface.mdx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  - `kex-0:7560300328640998700 -> 0xe494e36164e0a760140f3a9ab7dfdad38edac698f93d5239655dbd7499194760`
+- `did:iota:DSvXWs7FUch9MQcaUKmrRFZyHYcHwt3t3pbjvKsQBfep`
+  - `sign-0:16843234495045965331 -> 0xc6f0dbacd56156ff4c383d549ac61ada87f8aa69454f3bfae99f5fa9e093a5c3`
+  - `kex-0:16546298247591944074 -> 0x8e1d037cd343f84276ab737b638da9095bcb6052f7fd9628d21d20f434f9959a`
+  - `key:8559754420653090937 -> 0x4ef484a54aa16503878aa1ecaa6d73cb8254aefa3f80a569ed33ca685289d01e`
+
+Note how fragments (such as `kex-0`) can appear more than once, but the hash of the public key - calculated from the stored private key - makes the location unique in general. Although unlikely in practice, even the same private key can be used across different DIDs, which produces the same key location (here `sign-0:16843234495045965331`) and it's important that these are stored independently, so that deleting one does not accidentally delete the other. Hence why a key's full identifier in storage needs to be based on the DID _and_ the key location. That said, the following flattened structure also satisfies the requirements:
```
Comment (JelleMillenaar) [inline documentation/docs/advanced/storage_interface.mdx:53]: I feel like we really should provide some kind of :::caution here on the importance of this. We don't want people to overlook it and the way your wrote it makes it seem like the above is all intended and well.

File: documentation/docs/advanced/storage_interface.mdx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+## Indexing
+
+The interface has two methods called `did_list` and `did_exists`. These return the list of stored DIDs, and whether a DID exists in storage, respectively. Implementations are thus expected to maintain a list or index of stored DIDs. An identity created with `did_create` is added to the index, while an identity deleted through `did_purge` is removed from the index.
+
+If the storage implementation can be accessed concurrently, then access to the index needs to be synchronized, since it is unique per storage instance.
+
+## Implementation
+
+The identity.rs framework ships two implementations of `Storage`. The `MemStore` is an insecure in-memory implementation intended as an example implementation and for testing. The secure and recommended `Storage` is `Stronghold`, however `Stronghold` may also be interesting for implementers to look at, as it needs to deal with some challenges the in-memory version does not have.
```
Comment (JelleMillenaar) [inline documentation/docs/advanced/storage_interface.mdx:71]: We generally call it IOTA Identity framework

File: documentation/docs/advanced/storage_interface.mdx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+1. Generate the key at a random location, then derive the actual location and move the key there
+2. If moving a key is not possible, then an additional mapping from key location to some storage-internal location identifier can be maintained. Then it's possible to generate the key at some storage-internal location, calculate the key location and store the mapping.
+
+Since this also needs to happen before the DID can be derived from the public key, similar approaches can be used to work around the not-yet available DID partition key. Storages may choose to have one statically identified partition where keys are generated initially, and then moved from there. Storages whose restrictions do not allow for this, may want to use the flattened storage layout described in [example layout](#example-layout) and use the mapping approach.
+
+### Storage Test Suite
+
+The `StorageTestSuite` can be used to test the basic functionality of storage implementations. See its documentation for more details.
```
Comment (JelleMillenaar) [inline documentation/docs/advanced/storage_interface.mdx:86]: Can we link or does that require Rust API online? Can we already link JS?

**Review by JelleMillenaar — APPROVED** [review]
Excellent work! I am happy to merge as is, but made some minor comments. I also want to share one idea that you can judge if you like it or not: Should we add a simple list of all storage interface functions to be implement to provide an overview somewhere at the start? This can just be names with a simple one sentence explanation what it is supposed to do. I feel like that grounds the readers as to what this thing is.

File: documentation/docs/advanced/storage_interface.mdx (hunk 3 above)
Comment (PhilippGackstatter) [reply documentation/docs/advanced/storage_interface.mdx:86]: JS yes, but Rust is not on docs.rs yet, so no place to link to, unfortunately.

Comment (PhilippGackstatter) [reply documentation/docs/advanced/storage_interface.mdx:86]: Ah, actually the test suite is a post 0.5 feature, so I will have to remove this section here and add it later for the `dev` or 0.6 docs.

**Comment by PhilippGackstatter** [comment]
@JelleMillenaar Addressed all your comments and added a brief overview at the beginning, I think that's a good idea.

**Pull request merged**